_posts/2019-03-26-gloo-onion-layers.md

```diff
@@ -0,0 +1,250 @@
+// use gloo_timers::callback::Timeout;
+
+// Already, much nicer!
+let timeout = Timeout::new(500, move || {
+    do_some_operation();
+});
+
+// If we ever decide we want to cancel our delayed operation, all we do is drop
+// the `timeout` now:
+drop(timeout);
+
+// Or if we never want to cancel, we can use `forget`:
+timeout.forget();
+```
+
+## Layering on Futures and Streams
+
+The next layer to add is integrating with popular traits and libraries in the
+Rust ecosystem, like `Future`s or `serde`. For our running `gloo::timers`
+example, this means we implement a `Future` backed by `setTimeout`, and a
+`Stream` implementation backed by `setInterval`.
+
+```rust
+use futures::prelude::*;
+use gloo::timers::future::TimeoutFuture;
+
+// By using futures, we can use all the future combinator methods to build up a
+// description of some asynchronous task.
+let my_future = TimeoutFuture::new(500)
+    .and_then(|_| {
+        // Do some operation after 500 milliseconds...
+        do_some_operation();
+
+        // and then wait another 500 milliseconds...
+        TimeoutFuture::new(500)
+    })
+    .map(|_| {
+        // after which we do another operation!
+        do_another_operation();
+    })
+    .map_err(|err| {
+        handle_error(err);
+    });
+
+// Spawn our future to run it!
+wasm_bindgen_futures::spawn_local(my_future);
+```
+
+Note that we use `futures` 0.1 for now, because we've fought tooth and nail to
+get the Wasm ecosystem on stable Rust, but as soon as the new
+`std::future::Future` design is stable, we plan to switch over. We are very
+excited for `async`/`await` as well!
+
+## More Layers?
+
+That's all the layers we have for the `setTimeout` and `setInterval`
+APIs. Different Web APIs will have different sets of layers, and this is
+fine. Not every Web API uses callbacks, so it doesn't make sense to always have
+a `callback` module in every Gloo crate. The important part is that we are
+actively identifying layers, making them public and reusable, and building
+higher-level layers on top of lower-level layers.
+
+We will likely add even higher-level layers to other Web APIs where it makes
+sense. For example, the [File API][]'s `FileReader` interface exposes methods
+that you shouldn't call until after certain events have fired, and any attempt
+to call them earlier will throw. We can codify this as [a state machine-based
+`Future`][state-machine-future], that doesn't even give you the ability to call
+those methods until after the relevant events have fired and the state machine
+reaches a certain state. Leveraging types at compile time for ergonomics and
+correctness!
+
+Another future direction is adding more integration layers with more parts of
+the larger Rust crates ecosystem. For example, adding functional reactive
+programming-style layers via [the `futures-signals`
+crate][integrate-futures-signals] that the [`dominator`][dominator] framework is
```

Comment:
Obviously I'm not against the free advertising, but I'm not sure it makes sense to mention `dominator` here, it feels a little out of place.

Maybe it just needs to be reworded a bit? Like:

> via the `futures-signals` crate, which is also used by the `dominator` framework.